### PR TITLE
Handle rotation for primitives & camera

### PR DESCRIPTION
**Pull request by AxelF44**
Hello dear mates,

Here is my job about the rotation of camera and primitives.

All my brief have a Point and a Maj so Mister Tranchet/Valgrind you can't catch me on this point !

Here you can observe a rotation of a cylinders about 45 degrees on Z axis:
image

and without rotation:
image

Of course the light i up to the camera that mean that the cylinder receive light on his upside !

### CODE REVIEW

File: src/Core/main.cpp [hunk 1]
```diff
@@ -29,7 +29,7 @@ int main(int argc, char **argv)
     try {
         Raytracer::Scene scene(argv[1]);
         Raytracer::Renderer renderer(scene);
-        renderer.renderScene();
+        // renderer.renderScene();
```
Comment (Njord201) [inline src/Core/main.cpp:32]: Remove the comment or the line

File: include/Math/Vector3D.hpp [hunk 2]
```diff
@@ -82,6 +82,27 @@ namespace Math {
              */
             void translate(const Vector3D& ptr);
 
+            /**
+             * @brief Rotate a vector with an angle.
```
Comment (mathieurobert1) [inline include/Math/Vector3D.hpp:88]:
```suggestion
             * @brief Rotate a vector with an angle on the axis Z.
```

**Review by Njord201 — CHANGES_REQUESTED** [review]

File: include/Math/Vector3D.hpp [hunk 3]
```diff
@@ -82,6 +82,27 @@ namespace Math {
...
+             * 
+             * @param degrees how many to rotate
+             */
+            void rotateZ(double degrees);
+
+            /**
+             * @brief Rotate a vector with an angle.
```
Comment (mathieurobert1) [inline include/Math/Vector3D.hpp:95]:
```suggestion
             * @brief Rotate a vector with an angle on the axis Y.
```

File: include/Math/Vector3D.hpp [hunk 4]
```diff
@@ -82,6 +82,27 @@ namespace Math {
...
+             * 
+             * @param degrees how many to rotate
+             */
+            void rotateY(double degrees);
+
+            /**
+             * @brief Rotate a vector with an angle.
```
Comment (mathieurobert1) [inline include/Math/Vector3D.hpp:102]:
```suggestion
             * @brief Rotate a vector with an angle on the axis X.
```

File: src/Core/Parser/Scene.cpp [hunk 5]
```diff
@@ -212,6 +220,14 @@ int Raytracer::Scene::_parsePrimitiveSetting(const libconfig::Setting &primitive
             } else {
                 throw ParserException("Wrong Axis for Cylinder");
             }
+            if (cylinderArray[index].exists("rotation")) {
```
Comment (mathieurobert1) [inline src/Core/Parser/Scene.cpp:223]: add a line before the if statement as the others

File: src/Core/Parser/Scene.cpp [hunk 6]
```diff
@@ -212,6 +220,14 @@ int Raytracer::Scene::_parsePrimitiveSetting(const libconfig::Setting &primitive
...
+                libconfig::Setting& rotationSetting = cylinderArray[index].lookup("rotation");
+                const libconfig::Setting &rotationX = rotationSetting["x"];
+                const libconfig::Setting &rotationY = rotationSetting["y"];
+                const libconfig::Setting &rotationZ = rotationSetting["z"];
+                Math::Vector3D rotation(_parseValue(rotationX), _parseValue(rotationY), _parseValue(rotationZ));
+                newCylinder->setRotation(rotation);
+            }
```
Comment (mathieurobert1) [inline src/Core/Parser/Scene.cpp:230]: add a line after the if statement as the others

File: src/Math/Vector3D.cpp [hunk 7]
```diff
@@ -49,6 +49,52 @@ void Math::Vector3D::translate(const Vector3D& ptr)
...
+ 
+    double x = this->_vector[0];
+    double y = this->_vector[1];
+    double z = this->_vector[2];
+
+    this->_vector[0] = z * sin_theta + x * cos_theta;
+    this->_vector[1] = y;
+    this->_vector[2] = z * cos_theta - x * sin_theta;
+}
+
```
Comment (mathieurobert1) [inline src/Math/Vector3D.cpp:81]: useless line
```suggestion
```

File: src/Plugins/Primitives/Cylinder/Cylinder.cpp [hunk 8]
```diff
@@ -74,10 +87,18 @@ Math::Point3D Primitive::Cylinder::hitPoint(const Raytracer::Ray& ray) const
     Math::Vector3D rayOriginToHit = hitPoint - rayOrigin;
     if (IS_INVERSE(rayOriginToHit.x(), rayDirection.x()) || IS_INVERSE(rayOriginToHit.y(), rayDirection.y()) || IS_INVERSE(rayOriginToHit.z(), rayDirection.z()))
         return Math::Point3D(-1,-1,-1);
-
+    hitPoint.rotateX(this->_rotation.x());
+    hitPoint.rotateY(this->_rotation.y());
+    hitPoint.rotateZ(this->_rotation.z());
     return hitPoint;
 }
```
Comment (mathieurobert1) [inline src/Plugins/Primitives/Cylinder/Cylinder.cpp:95]: useless line
```suggestion
```

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
Some request before merging; but a very good job !
Well done Mr Fradet !

File: include/Primitives/Cylinder.hpp [hunk 9]
```diff
@@ -71,6 +71,13 @@ namespace Primitive {
              */
             void setAxis(const Primitive::Axis &axis);
 
+            /**
+             * @brief Set the Rotation object.
+             * 
+             * @param rotation rotation of cone
```
Comment (theophile-jr) [inline include/Primitives/Cylinder.hpp:77]: cylinder

**Review by Njord201 — DISMISSED** [review]

File: include/Primitives/Plane.hpp [hunk 10]
```diff
@@ -60,6 +60,13 @@ namespace Primitive {
              */
             void setAxis(const Primitive::Axis &axis);
 
+            /**
+             * @brief Set the Rotation object.
+             * 
+             * @param rotation rotation of cone
```
Comment (theophile-jr) [inline include/Primitives/Plane.hpp:66]: plane

File: include/Primitives/Sphere.hpp [hunk 11]
```diff
@@ -68,6 +68,13 @@ namespace Primitive {
              */
             void setMaterial(std::shared_ptr<Material::IMaterial> material);
 
+            /**
+             * @brief Set the Rotation object.
+             * 
+             * @param rotation rotation of cone
```
Comment (theophile-jr) [inline include/Primitives/Sphere.hpp:74]: sphere

**Review by theophile-jr — CHANGES_REQUESTED** [review]

**Review by theophile-jr — DISMISSED** [review]

File: include/Math/Vector3D.hpp [hunk 12]
```diff
@@ -84,6 +84,27 @@ namespace Math {
              */
             void translate(const Vector3D& ptr);
 
+            /**
+             * @brief Rotate a vector with an angle on the axis Z.
+             * 
+             * @param degrees how many to rotate
```
Comment (Njord201) [inline include/Math/Vector3D.hpp:90]:
```suggestion
             * @param degrees - Rotation value
```

File: include/Math/Vector3D.hpp [hunk 13]
```diff
@@ -84,6 +84,27 @@ namespace Math {
...
+             */
+            void rotateZ(double degrees);
+
+            /**
+             * @brief Rotate a vector with an angle on the axis Y.
+             * 
+             * @param degrees how many to rotate
```
Comment (Njord201) [inline include/Math/Vector3D.hpp:97]:
```suggestion
             * @param degrees - Rotation value
```

File: include/Math/Vector3D.hpp [hunk 14]
```diff
@@ -84,6 +84,27 @@ namespace Math {
...
+             */
+            void rotateY(double degrees);
+
+            /**
+             * @brief Rotate a vector with an angle on the axis X.
+             * 
+             * @param degrees how many to rotate
```
Comment (Njord201) [inline include/Math/Vector3D.hpp:104]:
```suggestion
             * @param degrees - Rotation value
```

File: include/Primitives/Cylinder.hpp [hunk 15]
```diff
@@ -71,6 +71,13 @@ namespace Primitive {
              */
             void setAxis(const Primitive::Axis &axis);
 
+            /**
+             * @brief Set the Rotation object.
+             * 
+             * @param rotation rotation of cylinder
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:77]:
```suggestion
             * @param rotation - Rotation value
```

File: include/Primitives/Plane.hpp [hunk 16]
```diff
@@ -60,6 +60,13 @@ namespace Primitive {
              */
             void setAxis(const Primitive::Axis &axis);
 
+            /**
+             * @brief Set the Rotation object.
+             * 
+             * @param rotation rotation of plane
```
Comment (Njord201) [inline include/Primitives/Plane.hpp:66]:
```suggestion
             * @param rotation - Rotation value
```

File: include/Primitives/Sphere.hpp [hunk 17]
```diff
@@ -68,6 +68,13 @@ namespace Primitive {
              */
             void setMaterial(std::shared_ptr<Material::IMaterial> material);
 
+            /**
+             * @brief Set the Rotation object.
+             * 
+             * @param rotation rotation of sphere
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:74]:
```suggestion
             * @param rotation - Rotation value
```

**Review by mathieurobert1 — DISMISSED** [review]
Nice work !

**Review by Njord201 — COMMENTED** [review]
To be discussed

**Review by theophile-jr — APPROVED** [review]

**Review by mathieurobert1 — APPROVED** [review]

**Review by Njord201 — COMMENTED** [review]
All's good

**Pull request merged**